### ISSUE TITLE
Notification Levels - Prioritise notifications by importance

### ISSUE THREAD

**Issue by Luehrsen** [WordPress/wp-feature-notifications#55]
Our attention spans are limited. And not every notification is equally important. So it would be tragic if an important notification would be buried under an otherwise unimportant bunch of marketing messages. 

I therefore suggest a 'level' system for notifications. Such a system could help group and prioritise notifications, so that users get access to information most important for them at the time. 

For example a SysOp could chose to ignore all notifications below a certain level to just consume information relevant to her or his task.

As for format, I suggest an integer based system, that classifies the levels like this:

| Level |     Title     |                                                                    Description                                                                    |
|:-----:|:-------------:|:-------------------------------------------------------------------------------------------------------------------------------------------------:|
|   0   |   Emergency   | A "panic" condition for when certain areas of WordPress stop working. Should be fixed immediately. Maybe triggered by the WP_Fatal_Error_Handler. |
|   1   |     Error     | Non-urgent failures that should be relayed to developers or admins. Examples could be issues from security plugins like WordFence.                |
|   2   |    Warning    | Not an error, but something might not be working as intended. An example of this is the Yoast SEO no-robots warning.                              |
|   3   |     Notice    | Noteworthy events in the systems, that are not errors. Example could be 'Post XY has been published'.                                             |
|   4   | Informational | Messages not directly related to the current WordPress. Marketing messages are here.                                                              |

This set of levels is inspired by [syslog](https://en.wikipedia.org/wiki/Syslog). I left everything out that usually goes into debug.log or tools like QueryMonitor.

**Comment by folletto** [comment]
I think theoretically this makes a lot of sense, but I've also a couple of reservations.

1. Syslog seems a very low level, technical tool to be used as a reference point. Is there a more consumer-facing standard we could take inspiration from?
2. How is this going to play with potential abuse? Given this isn't a controlled system, but anyone can write a plugin, and anyone can assign a priority level, I think we'll end up with "Emergency" used by a not-nice plugin trying to make a sale for example. 

I feel the logic works within a system where everyone plays for mutual benefit, but I've the impression in this case the potential abuse is really high.

When I look about large scale notification systems (i.e. Android, iOS) it seems they have all notifications at the same level, and handle instead grouping at source level — i.e. which app is triggering the specific notification. 

This seems a more viable approach, because once grouping happens by-source then the abuse potential is reduced to minimum: if an app sends 100 alerts, it's their own issue in burying their own important messages, compared to a system-wide prioritization system where everyone will send all notifications as "Emergency". 

For this reason, I'd lean to not include prioritization levels in this project, at least not for the initial release. 
I'm however happy to hear different opinions if there are ways to manage such abuse pattern that I've missed.

**Comment by Luehrsen** [comment]
> Syslog seems a very low level, technical tool to be used as a reference point. Is there a more consumer-facing standard we could take inspiration from?

From my point of view wp-notify is not (only) a consumer-facing tool. It is deployed on WP Admin, where a lot of roles with overlapping, but different relevancies towards certain information work on the same system. A lot of WordPress systems are maintained by huge teams of people that have different tasks and do not have time to give attention to every notification.

> When I look about large scale notification systems (i.e. Android, iOS) it seems they have all notifications at the same level, and handle instead grouping at source level — i.e. which app is triggering the specific notification.

Those systems, especially iOS, impose a set of guidelines on notifications being sent far stricter compared to what we could ever enforce. Additionally, even Apple's interface guidelines recommend giving the user control about what notifications to receive. (see https://developer.apple.com/design/human-interface-guidelines/ios/system-capabilities/notifications/#sending-marketing-notifications)

> I feel the logic works within a system where everyone plays for mutual benefit, but I've the impression in this case the potential abuse is really high.

While the potential for abuse is high, it is certainly lower than without such a system in which every notification would be equally as important. An enforcement of such rules must either be at the reputation or rules level. An option would be the adaptation of [Plugins Guideline 11 "... should not hijack the admin dashboard"](https://developer.wordpress.org/plugins/wordpress-org/detailed-plugin-guidelines/#11-plugins-should-not-hijack-the-admin-dashboard).

But then again: A system where everything is an emergency would equal a system we would have without prioritisation. Considering that abuse is coming from a small minority of vendors I would assume that the relevance for users would be higher than without the system.

**Comment by folletto** [comment]
> From my point of view wp-notify is not (only) a consumer-facing tool. It is deployed WP Admin, where a lot of roles with overlapping, but different relevancies towards certain information work on the same system. A lot of WordPress systems are maintained by huge teams of people that have different tasks and do not have time to give attention to every notification.

Note that the plan is to deliver per-user category, so not all people will get all notifications. That's already a way to segment that.

> Those systems, especially iOS, impose a set of guidelines on notifications being sent far stricter compared to what we could ever enforce.

Yes exactly. And even with that ability to restrict, they have no prioritization right? To me that's telling.

> An option would be the adaptation of Plugins Guideline 11 "... should not hijack the admin dashboard".

I like this idea. :)
In this case we could add a "Report this plugin" action maybe.

> But then again: A system where everything is an emergency would equal a system we would have without prioritisation. Considering that abuse is coming from a small minority of vendors I would assume that the relevance for users would be higher than without the system.

Agreed. I feel this will need to be considered when we design for this, the complexity of such a system is much higher, and given this statement, we might not over-highlight "Emergency" notifications, and instead play more with filters and settings. It's something to explore.

Thanks for the reply. Let's see what others think :)

**Comment by raaaahman** [comment]
This feature makes a lot of sense IMO, especially when WordPress' admin notices already implement it ( different highlights for Errors / Warnings / Notices ). So we should at least implement different highlights for notices in order to not have a feature regression here.

I also believe that regulation will be enforced by the users, uninstalling plugins that try to abuse the system.

**Comment by johnhooks** [comment]
> This seems a more viable approach, because once grouping happens by-source then the abuse potential is reduced to minimum: if an app sends 100 alerts, it's their own issue in burying their own important messages, compared to a system-wide prioritization system where everyone will send all notifications as "Emergency".

@folletto do I understand this right?

If priority is on a "source" level, a plugin's emergency priority shouldn't be given any preference in the general notifications display, unless a user specifically applies a filter. A plugin can use this system to display their own notifications in any order they prefer in their own settings page.

**Comment by folletto** [comment]
Sorry took me a bit to reconnect the neurons about this specific discussion :D 

What I was mentioning before is that if we are considering a consumer-level notification system like this, where we don't have control on the sources, it's really hard to build an effective functionality with "Notification Levels" because effectively the chance of abuse is really high.

This means:
1. As a baseline for v1.0, we can just ship without grouping
2. When we introduce in v1.x more advanced logic, we can include source grouping in the panel itself like Android and iOS do, where all the notifications from a single app get "stacked". This I believe being effective because by doing this now the app becomes very conscious that adding a second notification will effectively hide only _their_ other notifications, not other apps ones. This is a very thoroughly tested approach on existing system-wide notification systems so I think we can adapt this approach to WP too.
3. We could _then_ consider adding "Notification Levels" in a later v2.x. While I'd still consider this easily abused, if implemented AFTER the logic of grouping by source and used mostly for filtering (not to be visualized radically different as you stay, or get in front of other apps) then I think it can have some utility (i.e. I imagine an admin just filtering out errors and see what's going on).

Does it make sense?

**Comment by johnhooks** [comment]
@folletto thank you for the release break down.

I have been thinking about controlling abuse. Plugins are not sandboxes from the WordPress internals at all, the whole system relies on plugins respecting the user and their website. **We can and should develop the notification documentation and API to guide plugins to be conscientious and not overwhelm users.** Ultimately though, we can not stop a plugin from abuse, they have total access to the database and could add a notification that looks official and sourced from "WordPress", or for that matter sourced from another plugin.

**Comment by folletto** [comment]
I agree, while acknowledging all abuse can't be stopped, we must design a system that isn't easy to game in its own architecture. :)

**Comment by johnhooks** [comment]
I was spending too much time thinking how to "authorize" a plugin as a source, and realized it was a moot point.

**Comment by folletto** [comment]
Hah yes :D
